core/src/main/scala/org/ensime/indexer/SearchService.scala

```diff
@@ -95,54 +92,61 @@ class SearchService(
       }
     }.toSet ++ config.javaLibs.map(vfs.vfile)
 
-    // start indexing after all deletes have completed (not pretty)
-    val indexing = removed.map { _ =>
-      // could potentially do the db lookup in parallel
-      bases.filter(db.outOfDate).toList map {
-        case classfile if classfile.getName.getExtension == "class" => Future[Unit] {
-          val check = FileCheck(classfile)
-          val symbols = extractSymbols(classfile, classfile)
-          persist(check, symbols)
-        }(workerEC)
-
-        case jar => Future[Unit] {
-          try {
-            log.debug(s"indexing $jar")
-            val check = FileCheck(jar)
-            val symbols = scan(vfs.vjar(jar)) flatMap (extractSymbols(jar, _))
+    // individual DELETEs in H2 are really slow
+    for {
+      removed <- stale
+      _ <- Future.sequence(
+        removed.grouped(1000).map(delete)
+      )
+
+      // start indexing after all deletes have completed (not pretty)
+
+      basesWithOutOfDateInfo <- Future.sequence(bases.map(b => db.outOfDate(b).map((b, _))))
+      persisted <- Future.sequence(
+        basesWithOutOfDateInfo.collect { // is there a simpler way to do this, rather than collect after sequence?
+          case (base, outOfDate) if outOfDate => base
+        }.map {
+          case classfile if classfile.getName.getExtension == "class" => Future[Future[Unit]] {
+            val check = FileCheck(classfile)
+            val symbols = extractSymbols(classfile, classfile)
             persist(check, symbols)
-          } catch {
-            case e: Exception =>
-              log.error(s"Failed to index $jar", e)
-          }
-        }(workerEC)
+          }.flatMap(identity)
+
+          case jar => Future[Future[Unit]] {
+            try {
+              log.debug(s"indexing $jar")
+              val check = FileCheck(jar)
+              val symbols = scan(vfs.vjar(jar)) flatMap (extractSymbols(jar, _))
+              persist(check, symbols)
+            } catch {
+              case e: Exception =>
+                log.error(s"Failed to index $jar", e)
+                Future.successful(())
+            }
+          }.flatMap(identity)
+        }
+      )
+
+      _ = {
+        // delayed commits speedup initial indexing time
+        log.debug("committing index to disk...")
+        index.commit()
+        log.debug("...done committing index")
       }
-    }
 
-    val indexed = indexing.flatMap { w => Future.sequence(w) }.map(_.size)
-    val indexedAndCommitted = indexed map { count =>
-      // delayed commits speedup initial indexing time
-      log.debug("committing index to disk...")
-      index.commit()
-      log.debug("...done committing index")
-      count
-    }
-
-    for {
-      r <- removed
-      i <- indexedAndCommitted
-    } yield (r, i)
+    } yield (removed.size, persisted.size)
```

Comment:
a single comprehension encompassing all of the above seemed cleaner here and since I had to touch it anyways